### PR TITLE
Human-friendly durations in PromQL

### PR DESCRIPTION
**Pull request by roidelapluie**
This is the PromQL / documentation part of https://github.com/prometheus/common/pull/246

### CODE REVIEW

**Comment by roidelapluie** [comment]
Obviously tests will fail until we release common. This is however ready for review.

**Comment by brian-brazil** [comment]
model.Duratoin is used in far more places than this. Flag and config docs need updating too, as will AM.

File: docs/configuration/configuration.md [hunk 1]
```diff
@@ -32,7 +32,7 @@ value is set to the specified default.
 Generic placeholders are defined as follows:
 
 * `<boolean>`: a boolean that can take the values `true` or `false`
-* `<duration>`: a duration matching the regular expression `[0-9]+(ms|[smhdwy])`
+* `<duration>`: a duration matching the regular expression `(([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?`
```
Comment (brian-brazil) [inline docs/configuration/configuration.md:35]: This doesn't cover 0.

This is also beyond what a human can easily eyeball, so I'd include an example or two.

File: docs/querying/basics.md [hunk 2]
```diff
@@ -160,11 +168,15 @@ following units:
...
 * `y` - years
 
-In this example, we select all the values we have recorded within the last 5
-minutes for all time series that have the metric name `http_requests_total` and
-a `job` label set to `prometheus`:
+Time durations can be combined, by concatenation. Units must be ordered from the
+longest to the shortest. A given unit must only appear once in a time duration.
 
-    http_requests_total{job="prometheus"}[5m]
+Here are some exemples of valid time durations:
```
Comment (brian-brazil) [inline docs/querying/basics.md:174]: examples

File: docs/querying/basics.md [hunk 3]
```diff
@@ -160,11 +168,15 @@ following units:
 * `w` - weeks
 * `y` - years
```
Comment (brian-brazil) [inline docs/querying/basics.md:169]: This should also mention that years are always 365 days

File: promql/parser/parse_test.go [hunk 4]
```diff
@@ -1638,14 +1668,6 @@ var testExpr = []struct {
 		input:  `foo[0m]`,
 		fail:   true,
 		errMsg: "duration must be greater than 0",
-	}, {
-		input:  `foo[5m30s]`,
```
Comment (brian-brazil) [inline promql/parser/parse_test.go:1642]: Some new negative tests would be good, like 1hm, 1m1, 1m2h, 1m1m.

File: promql/parser/lex.go [hunk 5]
```diff
@@ -701,17 +697,32 @@ func lexNumberOrDuration(l *Lexer) stateFn {
...
+	if !l.accept("smhdwy") {
+		return false
+	}
+	// Next char can be another number then a unit.
+	for l.accept("0123456789") {
+		for l.accept("0123456789") {
+		}
+		// y is no longer in the list as it should always come first in
+		// durations.
+		if !l.accept("smhdw") {
```
Comment (brian-brazil) [inline promql/parser/lex.go:719]: Does this cover ms?

Comment (roidelapluie) [reply promql/parser/lex.go:719]: ms is not supported in PromQL: http://demo.robustperception.io:9090/graph?g0.range_input=1h&g0.expr=up%5B5ms%5D&g0.tab=1

should I fix that?

Comment (brian-brazil) [reply promql/parser/lex.go:719]: For consistency we probably should, even if it's not much use.

File: docs/configuration/configuration.md (hunk 1 above)
Comment (roidelapluie) [reply docs/configuration/configuration.md:35]: In which case is 0 valid in the context of this file?

Comment (brian-brazil) [reply docs/configuration/configuration.md:35]: it's a model.Duration, and they take 0 - whether it's a flag or a config field.

Comment (roidelapluie) [reply docs/configuration/configuration.md]: done

File: docs/querying/basics.md (hunk 2 above)
Comment (roidelapluie) [reply docs/querying/basics.md]: addressed

File: docs/querying/basics.md (hunk 3 above)
Comment (roidelapluie) [reply docs/querying/basics.md]: done, same for  days and weeks. That is what the comments on time.Duration also specify:
https://github.com/prometheus/common/blob/217fd62d44a79db8f842113234ae9383b8374983/model/time.go#L186-L187

File: promql/parser/parse_test.go (hunk 4 above)
Comment (roidelapluie) [reply promql/parser/parse_test.go:1642]: added negative tests

File: promql/parser/lex.go (hunk 5 above)
Comment (roidelapluie) [reply promql/parser/lex.go:722]: Added support for ms

**Comment by brian-brazil** [comment]
:+1:

**Pull request merged**